File: 346. Moving Average from Data Stream.py
class MovingAverage:

    def __init__(self, size: int):
        """
        Initialize your data structure here.
        """
        self.q = collections.deque()
        self.size = size
        self.window_sum = 0
        self.count = 0

    def next(self, val: int) -> float:
        if self.count < self.size:
            self.q.append(val)
            self.window_sum += val
            self.count += 1
        else:
            toremove = self.q.popleft()
            self.q.append(val)
            self.window_sum += val - toremove
        return sum(self.q) / self.count


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)

File: 346. Moving Average from Data Stream.java
class MovingAverage {
    private int size;
    private int currsize;
    private int sum;
    private Deque<Integer> cache;

    /** Initialize your data structure here. */
    public MovingAverage(int size) {
        this.size = size;
        currsize = 0;
        sum = 0;
        cache = new ArrayDeque<>(size);
    }
    
    public double next(int val) {
        if (currsize == size) {
            sum -= cache.poll();
            sum += val;
            cache.offer(val);
        } else {
            sum += val;
            cache.offer(val);
            currsize++;
        }
        return (double) sum / (double) currsize;
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */